src/leveled_sst.erl

```diff
@@ -680,39 +669,43 @@ starting(complete_l0startup, State) ->
             blockindex_cache = BlockIndex,
             high_modified_date = HighModDate,
             monitor = Monitor}};
-starting({sst_returnslot, FetchedSlot, FetchFun, SlotCount}, State) ->
-    Self = self(),
-    FetchedSlots = 
+starting(cast, {sst_returnslot, FetchedSlot, FetchFun, SlotCount}, State) ->
+    FetchedSlots =
         case FetchedSlot of
             none ->
                 [];
             _ ->
                 [FetchedSlot|State#state.new_slots]
-        end, 
+        end,
     case length(FetchedSlots) == SlotCount of
         true ->
-            gen_fsm:send_event(Self, complete_l0startup),
-            {next_state,
-                starting,
-                % Reverse the slots so that they are back in the expected
-                % order
-                State#state{new_slots = lists:reverse(FetchedSlots)}};
+            {keep_state,
+                %% Reverse the slots so that they are back in the expected
+                %% order
+                State#state{new_slots = lists:reverse(FetchedSlots)},
+            [{next_event, cast, complete_l0startup}]};
         false ->
+            Self = self(),
             ReturnFun =
                 fun(NextSlot) ->
-                    gen_fsm:send_event(Self, 
+                    gen_statem:cast(Self,
                                         {sst_returnslot, NextSlot,
                                             FetchFun, SlotCount})
                 end,
             FetchFun(length(FetchedSlots) + 1, ReturnFun),
-            {next_state,
-                starting,
+            {keep_state,
                 State#state{new_slots = FetchedSlots}}
-    end.
+    end;
+starting(cast, {update_blockindex_cache, BIC}, State) ->
+    handle_update_blockindex_cache(BIC, State);
 
```

Comment:
This message cannot be received in the startup state, as it is prompted from get_slots which can only occur in the reader state